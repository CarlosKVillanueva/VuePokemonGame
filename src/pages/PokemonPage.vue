<template>

	<h1 v-if="!pokemon">Espere por favor...</h1>
	<div v-else>
		<h1>Quien es este Pokemon?</h1>
		<PokemonImage
			:pokemon-id="pokemon.id"
			:show-pokemon="showPokemon"
			class="app-center"
		/>
		<PokemonOptions
			:pokemons="pokemonArr"
			@selected-pokemon="checkAnswer"
			class="app-center"
		/>
		<template v-if="showAnswer">
			<h2 class="fade-in" :class="{ acerto, error } ">{{ message }}</h2>
			<button @click="newGame">
				Nuevo Juego
			</button>
		</template>

	</div>
</template>

<script>
import PokemonImage from "@/components/PokemonImage";
import PokemonOptions from "@/components/PokemonOptions";
import getPokemonOptions  from "@/helpers/getPokemonOptions";

// console.log( getPokemonOptions() );

export default {
	components: { PokemonImage, PokemonOptions },
	data() {
		return {
			pokemonArr: [],
			pokemon: null,
			showPokemon: false,
			showAnswer: false,
			message: false,
			acerto: false,
			error: false
		};
	},
	methods: {
		async mixPokemonArr() {
			this.pokemonArr = await getPokemonOptions();
			const rand = Math.floor( Math.random() * 4 );
			this.pokemon = this.pokemonArr[ rand ];
			console.log(this.pokemon)
		},
		checkAnswer(pokemonId) {
			this.acerto = false;
			this.error = false;
			this.showPokemon = true;
			this.showAnswer = true;
			if ( pokemonId === this.pokemon.id ) {
				this.message = `Correcto, ${this.pokemon.name}!`
				this.acerto = true;
			} else {
				this.message = `Ooopss, it was ${this.pokemon.name}!`
				this.error = true;
			}
		},
		newGame() {
			this.acerto = false;
			this.error = false;
			this.showPokemon = false;
			this.showAnswer = false;
			this.pokemon = null;
			this.pokemonArr = [];
			this.mixPokemonArr();
		},
	},
	mounted() {
		this.mixPokemonArr();
	}
}
</script>

<style scoped>

</style>