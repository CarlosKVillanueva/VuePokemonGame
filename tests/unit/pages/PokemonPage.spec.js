import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from "@/pages/PokemonPage"
import { pokemonsMock } from "../mocks/pokemons.mock"



describe( 'PokemonPage', function () {
    let wrapper
    let mixPokemonArraySpy

    beforeEach(() => {
        mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, "mixPokemonArr" )
        wrapper = shallowMount( PokemonPage )
    })

    test( 'Must Call mixPokemonArray on mounted', () => {
        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    } );
    
    test('Must match the Snapshot when Pokemons been loaded', () => {
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    acerto: false,
                    error: false
                }
            }
        })
        expect( wrapper.html() ).toMatchSnapshot()
    });
    
    test('Must show the PokemonImage and PokemonOptions components', () => {


        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    acerto: false,
                    error: false
                }
            }
        })

        let imageStub = wrapper.find( 'pokemon-image-stub' )
        let optionsStub = wrapper.find( 'pokemon-options-stub' )

        expect( imageStub.exists() ).toBeTruthy()
        expect( optionsStub.exists() ).toBeTruthy()

        expect( imageStub.attributes('pokemonid') ).toBe( '1' )
        expect( optionsStub.attributes( 'pokemons' ) ).toBeTruthy()
    });

    test('CheckAnswer Test', async () => {
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[ 0 ],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    acerto: false,
                    error: false
                }
            }
        } )

        await wrapper.vm.checkAnswer( 1 )

        expect( wrapper.find( 'h2' ).exists() ).toBeTruthy()
        expect( wrapper.find( 'button' ).exists() ).toBeTruthy()

        expect( wrapper.vm.showPokemon ).toBeTruthy()
        expect( wrapper.vm.showAnswer ).toBeTruthy()
        expect( wrapper.find( 'h2' ).text() ).toBe( `Correcto, ${pokemonsMock[0].name}!` )

        await wrapper.vm.checkAnswer( 10 )
        expect( wrapper.vm.message ).toBe( `Ooopss, it was ${ pokemonsMock[ 0 ].name }!` )
    });

} );