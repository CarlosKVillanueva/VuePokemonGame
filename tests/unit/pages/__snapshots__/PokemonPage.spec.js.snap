// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Must match the Snapshot when Pokemons been loaded 1`] = `
<div>
  <h1>Quien es este Pokemon?</h1>
  <pokemon-image-stub pokemonid="1" showpokemon="false" class="app-center"></pokemon-image-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" class="app-center"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
